File: Medium/Java/Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal(105).java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        
        return helper(0, 0, inorder.length - 1, preorder, inorder);
    }

    public TreeNode helper(int preStart, int inStart, int inEnd, int[] preorder, int[] inorder) {
        
        // 1. Base case
        if (preStart > preorder.length || inStart > inEnd){
            return null;
        }
        
        // 2. Build the current node.
        TreeNode cur = new TreeNode(preorder[preStart]);
        
        // 3. Find the pivot in the inorder.
        int inPivot = 0;
        for (int i = 0; i < inorder.length; i++){
            if (inorder[i] == preorder[preStart]){
                inPivot = i;
                break;
            }
        }
        
        // 4. return root.
        cur.left  = helper(preStart + 1, inStart, inPivot - 1, preorder, inorder);
        // key
        cur.right = helper(preStart + inPivot - inStart + 1, inPivot + 1, inEnd, preorder, inorder);
        return cur;
    }
}

File: Medium/C++/Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal(105).cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution 
{
    unordered_map<int, int> m_in;

    TreeNode* f(vector<int>& pre, int i, int j, int J)
    {
        if(j>J) return nullptr;
        auto y{m_in[pre[i]]};
        return new TreeNode(pre[i], f(pre, i+1, j, y-1), f(pre, i+1+y-j, y+1, J));
    }
public:    
    TreeNode* buildTree(vector<int>& pre, vector<int>& in) 
    {
        for(auto i{0}; i<size(pre); m_in[in[i]]=i, ++i);
        return f(pre, 0, 0, size(in)-1);    
    }
};